codepropertygraph/src/main/scala/io/shiftleft/utils/TimeMetric.scala

```diff
@@ -0,0 +1,349 @@
+  // Stage wise performance will be only recorded if parallel recording is started through initialize()
+  val stagePerformance: mutable.Map[String, NumberProcessor] = mutable.Map()
+  private val start                                          = cal.getTime
+  private var subStageStartTime                              = start
+  private var newTime                                        = start
+  private var stageStartTime                                 = start
+  private val ACROSS_PROCESS_STAGE                           = "<ACROSS_ALL>"
+  private val STAGE_NOT_SET                                  = "<not set>"
+  private val STARTED                                        = "Started"
+  private val DONE                                           = "Done"
+  private var lastStageName                                  = STAGE_NOT_SET
+  private var lastSubStageName                               = STAGE_NOT_SET
+  private var additionalMetaData                             = STAGE_NOT_SET
+  private var alsoLogConsole                                 = false
+  private val logger: (String, Boolean) => Unit = (msg, subStage) => {
+    if (alsoLogConsole && (!subStage || !supressSubstages))
+      println(msg)
+    baseLogger.debug(msg)
+  }
+  private val baseLogger: Logger             = LoggerFactory.getLogger(getClass)
+  private var recorder: Option[TimeRecorder] = None
+  private var fullName: Boolean              = false
+  private var maxBuffers: MaxPrintBuffers    = MaxPrintBuffers()
+  private var supressSubstages: Boolean      = true
+  private lazy val maxStartingBuffer = {
+    if (fullName) {
+      maxBuffers.MAX_FULLNAME_BUFFER + maxBuffers.MAX_DONE_TIME_BUFFER
+    } else {
+      maxBuffers.MAX_NAME_BUFFER + maxBuffers.MAX_DONE_TIME_BUFFER
+    }
+  }
+  private lazy val maxEndingBuffer = {
+    if (fullName) {
+      maxBuffers.MAX_FULLNAME_BUFFER
+    } else {
+      maxBuffers.MAX_NAME_BUFFER
+    }
+  }
+  private val osBean = ManagementFactory.getOperatingSystemMXBean().asInstanceOf[OperatingSystemMXBean]
+
+  def initialize(
+    useFullName: Boolean = false,
+    consoleLog: Boolean = false,
+    timeMetricRecordConfig: Option[TimeMetricRecordConfig] = None,
+    maxPrintBuffers: MaxPrintBuffers = MaxPrintBuffers(),
+    supressSubStagesConsoleLog: Boolean = true
+  ): Unit = {
+    fullName = useFullName
+    maxBuffers = maxPrintBuffers
+    supressSubstages = supressSubStagesConsoleLog
+    alsoLogConsole = consoleLog
+    timeMetricRecordConfig match {
+      case Some(config) =>
+        stagePerformance += (ACROSS_PROCESS_STAGE -> NumberProcessor())
+        recorder = Some(TimeRecorder(config))
+        recorder.get.start()
+      case _ =>
+    }
+  }
+
+  def setSupressSubstagesFlag(flag: Boolean): Unit = supressSubstages = flag
+
+  private class TimeRecorder(timeMetricRecordConfig: TimeMetricRecordConfig) extends Thread {
+
+    val queue: util.Queue[String] = new ConcurrentLinkedQueue[String]()
```

Comment:
If `TimeMetric` becomes a class, would a recorder have multiple threads accessing the queue?

Reply:
As per my previous comment, Object will be the more suitable option for `TimeMetric`